akka-typed-tests/src/test/scala/akka/typed/scaladsl/persistence/ApiTest.scala

```diff
@@ -0,0 +1,394 @@
+package akka.typed.scaladsl.persistence
+
+import akka.typed
+
+import scala.concurrent.duration._
+import akka.typed.{ ActorRef, Behavior, ExtensibleBehavior, Signal, Terminated }
+import akka.typed.scaladsl.{ ActorContext, TimerScheduler }
+
+import scala.concurrent.ExecutionContext
+import scala.reflect.ClassTag
+
+class ApiTest {
+  object TypedPersistentActor {
+
+    sealed trait PersistentEffect[+Event]
+
+    case object PersistNothing extends PersistentEffect[Nothing]
+
+    case class Persist[Event](event: Event, callbacks: List[Event ⇒ Unit] = Nil) extends PersistentEffect[Event] {
+      def andThen(callback: Event ⇒ Unit) = Persist(event, callback :: callbacks)
+    }
```

Comment:
Same as I mentioned on the multi-dc API: https://github.com/typesafehub/akka-commercial-addons/pull/107#discussion_r138634348

All things must be able to andThen otherwise we make it hard for people to abstract "and do this thing always anyway".

Reply:
Hmm, I generally agree, though this did get kind of hairy (https://github.com/akka/akka/pull/23631/commits/ad11555c0d6c21ec36e9b2f6fe5f45521b20f1dc) - do you have a nicer pattern? I tried moving `callbacks` to `PersistentEffect` but that caused variance problems I wasn't sure how to get rid of.

Reply:
isn't that just implementation details, I solved it at the other place

Reply:
Yes, that's an implementation thing. Cleaned it up a bit, still not entirely happy with it - having everything inside the `ReplicatedEntity` abstract superclass does make the typing/generics a bit nicer, but that wouldn't be very 'akka-typed-like'...